### PR TITLE
Add X11 abilities to Kitematic

### LINKED ISSUE #270: Integrate x11
Just a simple setting to allow display settings be added during a container run statement, for x11 forwarding :smile:

### PR DESCRIPTION
**Pull request by koddsson**
This patch will add a new setting variable in the settings view to set a `DISPLAY` path for lunching of X11 containers such as `jess/chrome`.

On unix (untested) you simply fill out the input with "unix:0" (if your current display is 0) and start a X11 container. On Mac OSX you need start a `socat` daemon. Here's an example:

    socat TCP-LISTEN:6000,reuseaddr,fork UNIX-CLIENT:\"$DISPLAY\"

With a daemon running you can set the X11 display path to the virtualbox network interface that is running docker ("192.168.59.3:0").

When lunching a container we'll check for the settings file and inject into the container options if it exists.

Try running Chrome with `jess/chrome`!

Fixes #270.

### CODE REVIEW

**Comment by Elesant** [comment]
This is super cool @koddsson! Thanks for adding this. I'll give this a try :)

**Comment by koddsson** [comment]
Please do and tell me what you think @Elesant :)

**Comment by jeffdm** [comment]
Sorry for not having a look yet. A little swamped :sweat_smile: 

I like the looks of this on first glance. Great placement of it in preferences - since it's a rather _advanced_ feature.  Taking it for a spin later today!

**Comment by robtec** [comment]
@koddsson Pretty cool dude, nice implementation :+1:

**Comment by robtec** [comment]
@koddsson So I notice after setting the x11 display path, when you leave the preferences page and go back in the edit box is blank yet launched containers have the DISPLAY ENV arg set as the original value entered, maybe the preference window needs to check if the value is set and populate the edit box.. or maybe this is just happening to me :laughing: :sob:

**Comment by koddsson** [comment]
@robtec That's weird! This wasn't happening to me but I can reproduce this now! I'll see if I can fix it. :wrench:

**Comment by koddsson** [comment]
@robtec Should be fixed. This is hackathon code after all! :runner: 

Give it another go and tell me what you think!

**Comment by jeffdm** [comment]
Looking :smiley:

**Comment by jeffdm** [comment]
Couldn't quite get this working? What did I miss :)?

- `brew install socat`
- `socat TCP-LISTEN:6000,reuseaddr,fork UNIX-CLIENT:\"192.168.99.100\"`
- Add `192.168.99.100` to the Kitematic X11 preferences
- Run jess/chrome with Kitematic

The container just stops with:

```[1:1:0326/182328:ERROR:browser_main_loop.cc(170)] Running without the SUID sandbox! See https://code.google.com/p/chromium/wiki/LinuxSUIDSandboxDevelopment for more information on developing with the sandbox on.
[1:1:0326/182328:ERROR:browser_main_loop.cc(216)] Gtk: cannot open display: 192.168.99.100:0```

**Comment by koddsson** [comment]
socat TCP-LISTEN:6000,reuseaddr,fork UNIX-CLIENT:\"$DISPLAY\"

instead of

    socat TCP-LISTEN:6000,reuseaddr,fork UNIX-CLIENT:\"192.168.99.100\"

?

**Comment by jeffdm** [comment]
What should `echo $DISPLAY` output? Mine's just blank

**Comment by koddsson** [comment]
```
☁  ~  echo $DISPLAY
/private/tmp/com.apple.launchd.BA1bEF6iVr/org.macosforge.xquartz:0
```

I just remembered that I did need to reboot once I installed `socat` to get that variable set correctly. Not sure if there is a better way.

**Comment by jeffdm** [comment]
Silly me, I didn't have XQuartz installed

**Comment by koddsson** [comment]
I made a GIF! :cinema: 

kitematic2

**Comment by jeffdm** [comment]
Nice! Ok let me try again :smiley:

**Comment by jeffdm** [comment]
While that sets up, I noticed this feature has many dependencies on Mac:
- XQuartz
- socat
- user needs to start a socat process

I'm wondering, do you feel that this feature may be a bit hard to make use of on the Mac OS X version of Kitematic (as opposed to Linux version)?

**Comment by koddsson** [comment]
>I'm wondering, do you feel that this feature may be a bit hard to make use of on the Mac OS X version of Kitematic (as opposed to Linux version)?

Yeah it might be a bit much. Since it's essentially just injecting some environ variable maybe it would be better suited to add the ability to set those variables in kitematic and then getting X11 apps running is just a matter of some documentation?

**Comment by robtec** [comment]
@koddsson Nice fix, working great :+1: 

the `socat` and `XQuartz` dependencies are a pain :cry:

**Comment by jeffdm** [comment]
@koddsson Love that idea! Sorry I didn't mention the feedback sooner in the PR. I really appreciate you writing the code for this.

We currently support adding environment variables in the container settings pane so I agree it would be best to have docs with the steps:
- Download XQuartz
- Download socat
- Start the socat process
- Start Kitematic and create the `jess/chrome` container
- Change the `DISPLAY` environment variable for the newly created container

I think moving to docs makes a lot of sense since the steps are mostly manual. I'd love to include a new article with this Gif and steps in the kitematic docs (https://kitematic.com/docs). We'll be open sourcing the docs in the next week or so.

What do you think?

**Comment by koddsson** [comment]
Yeah I agree. 

Thanks for taking the time to go over this with me @jeffdm! I'll definitely make more contributions to Kitematic later on!

**Comment by jeffdm** [comment]
@koddsson Anytime! Sorry again for not getting to this sooner. Thanks so much for creating a PR!
